### PR TITLE
bip32: derive_xpriv should not return a Result

### PR DESCRIPTION
**Pull request by storopoli**
We discussed in #2752 that `derive_priv` never fails.

This PR addresses that issue.

### CODE REVIEW

**Comment by storopoli** [comment]
Should we tackle the `derive_pub` in this PR as well?

I draft something very simple like:

```rust
/// Public->Public child key derivation
    pub fn ckd_pub<C: secp256k1::Verification>(&self, secp: &Secp256k1<C>, i: ChildNumber) -> Xpub {
        let (sk, chain_code) = self.ckd_pub_tweak(i).expect("could not derive child key");
        let tweaked = self.public_key.add_exp_tweak(secp, &sk.into())?;

        // Check for invalid child number according to BIP32.
        // This means that if the derivation is hardened, we panic.

        match i {
            ChildNumber::Hardened { .. } => panic!("cannot derive hardened key from public key"),
            ChildNumber::Normal { .. } => Xpub {
                network: self.network,
                depth: self.depth + 1,
                parent_fingerprint: self.fingerprint(),
                child_number: i,
                public_key: tweaked,
                chain_code,
            },
        }
    }
```

**Comment by apoelstra** [comment]
> Should we tackle the `derive_pub` in this PR as well?

No, because it will require changing the `ChildNumber` structure to contain specific types rather than `u32`s, which is a much larger change. Then `ckd_pub` would accept the `HardenedChildNumber` type rather than the general `ChildNumber` enum. (If you want to tackle it, please do, but in another PR!)

Just replacing the error path with a panic is not the right approach :).

File: bitcoin/examples/ecdsa-psbt.rs [hunk 1]
```diff
@@ -130,7 +130,9 @@ impl ColdStorage {
     }
 
     /// Returns the fingerprint for the master extended public key.
-    fn master_fingerprint(&self) -> Fingerprint { self.master_xpub.fingerprint() }
+    fn master_fingerprint(&self) -> Fingerprint {
+        self.master_xpub.fingerprint()
+    }
```
Comment (tcharding) [inline bitcoin/examples/ecdsa-psbt.rs:135]: This is an unrelated (incorrect) formatting change. Have you got your editor set to format on save but using the stable toolchain?

Comment (tcharding) [reply bitcoin/examples/ecdsa-psbt.rs:135]: There's a bunch of other places, will review after those are gone. Sorry, `rustfmt` presents a bunch of basically unsolvable annoyances, one of which is it has to be run in this codebase with `nightly`.

Comment (storopoli) [reply bitcoin/examples/ecdsa-psbt.rs]: They are gone, thanks for pointing them out.

File: bitcoin/src/bip32.rs [hunk 2]
```diff
@@ -583,19 +583,21 @@ impl Xpriv {
             .expect("BIP32 internal private key representation is broken")
     }
 
-    /// Attempts to derive an extended private key from a path.
+    /// Derive an extended private key from a path.
```
Comment (tcharding) [inline bitcoin/src/bip32.rs:586]: We don't use the imperative tense for function docs. FYI when we discussed and started implementing this it was the correct choice, to the best of our knowledge, even though there are many crates that do use the imperative. Here is an arbitrary example from the std lib: 

https://doc.rust-lang.org/std/string/struct.String.html#method.reserve

I'm not super good with grammar, please correct me if you catch me making any mistakes, second language English speakers often have better grammar than first language English speakers in my experience, well often better than me anyway. I believe the tense we use is third person, where the function is the third person.

Comment (tcharding) [reply bitcoin/src/bip32.rs:586]: I wouldn't normally pick this up in review but since you are interested in docs I thought it worth mentioning.

Comment (storopoli) [reply bitcoin/src/bip32.rs:586]: Really helpful thanks! Yes, I am interested in docs and technical documentation.
I agree with this. Will update it.

**Comment by apoelstra** [comment]
7df90071c91316149e1c6e1ca984b649f68d6820 looks good except that `ckd_priv` should not return a result either. Then you can drop the new `expect`.

**Comment by junderw** [comment]
> [7df9007](https://github.com/rust-bitcoin/rust-bitcoin/commit/7df90071c91316149e1c6e1ca984b649f68d6820) looks good except that `ckd_priv` should not return a result either. Then you can drop the new `expect`.

Interestingly enough, all the Result-returning-calls within ckd_priv all expect using `statistically impossible to hit`...

**Comment by storopoli** [comment]
> [7df9007](https://github.com/rust-bitcoin/rust-bitcoin/commit/7df90071c91316149e1c6e1ca984b649f68d6820) looks good except that `ckd_priv` should not return a result either. Then you can drop the new `expect`.

Done in d353be45462e46a0d430224ae93aaa6d15488058

**Review by apoelstra — APPROVED** [review]
ACK d353be45462e46a0d430224ae93aaa6d15488058

**Review by tcharding — APPROVED** [review]
ACK d353be45462e46a0d430224ae93aaa6d15488058

**Review by sanket1729 — APPROVED** [review]
ACK d353be45462e46a0d430224ae93aaa6d15488058
